### DISCUSSION TITLE
Contributing to library (group time series split)

### DISCUSSION THREAD

**Discussion by labdmitriy** [rasbt/mlxtend#905] [Ideas]
Hi Sebastian,

Do you have any interest in extending your great library with group time series split, which combines several ideas from different libraries?
Initially I wrote it for my personal tasks, but after that I decided to publish it on GitHub (https://github.com/labdmitriy/ml-lab), wrote an article about that (https://medium.com/@labdmitriy/advanced-group-time-series-validation-bb00d4a74bcc), and provided Jupyter Notebook with examples (https://github.com/labdmitriy/ml-lab/blob/master/notebooks/GroupTimeSeriesSplit.ipynb).
It would be an honor for me to practice in contributing (I don't have any experience in it for open source project) espeсially to your library which I used frequently enough.

Thank you.

**Comment by rasbt** [comment]
Thanks for asking! To be honest, that'd be an awesome contribution! MLxtend is exactly for these sort of things that are beyond the scope of other tools like scikit-learn but are super useful in practice.

I am also super happy to guide you through the process.

The big, first question would be, should this go under mlxtend.evaluate, or should we create a new submodule mlxtend.time_series

What speaks for mlxtend.evaluate is that we have similar methods there such as BootstrapOutOfBag (http://rasbt.github.io/mlxtend/user_guide/evaluate/BootstrapOutOfBag/). On the other hand, what speaks against it is discoverability. 

What are your thoughts?

**Reply by labdmitriy** [reply]
Hi Sebastian, 

Great!
I am agree with you about the module mlxtend.evaluate, because I suppose that it can be used for any ordered data with continuous groups, not just for the time. But of course your choice will be final)
So I can explore your contribution guide for the package during the weekend and try to prepare initial version of the code step-by-step according to your guide. 
Or should we discuss in more detail first?

Thank you!

**Comment by rasbt** [comment]
Hm, so maybe we could make a version for `mlxtend.evaluate` and if there are more time series related things, we can think about moving it out later (with a deprecation warning). 

In this case, I recommend making a copy of https://github.com/rasbt/mlxtend/blob/master/mlxtend/evaluate/bootstrap_outofbag.py and going from there. I am happy to help along the way.

**Comment by labdmitriy** [comment]
Hi @rasbt ,

I wrote the article about contributing to open source based on experience of contributing to your `mlxtend` library:
https://labdmitriy.github.io/blog/contributing-in-open-source/

**Reply by rasbt** [reply]
Wohoo! That's exciting! Just downloaded it to my e-reader and will read it soon!!

**Reply by rasbt** [reply]
Hey @labdmitriy. I just read the article and wow, I had no idea it was your first open source contribution. To be honest, I thought you've been doing that for a long time since you knew all the details and best-practices. Next to the great contrib, I appreciated all the back and forth regarding the pre-commit hooks, and I think I learned more from you than the other way around. Anyways, great article and summary of how to contribute to open source projects. I will keep the article bookmarked and will recommend it to students who are new to everything and would like to know what the journey is like!

**Reply by labdmitriy** [reply]
Thanks a lot @rasbt 😅
I hope you also saw the embedded video, unfortunately it is working only on desktop but it is interesting.
My offer to do or write something together will be relevant in any case 🙂

P.S.
1. Also while I was writing the article I noticed that flake8 is in GitHub Action with testing and is executed before isort/black, so probably it will be better to move flake8 after it.
2. Recently I found the great book about scientific visualization using Python and matplotlib and today I wrote the article about how to build the latest version of the book from scratch, maybe it will be interesting for you or your students - https://labdmitriy.github.io/blog/building-scientific-visualization-book/

**Reply by rasbt** [reply]
The video looks cool. Yeah, it unfortunately doesn't play on Safari but works just fine on Chrome. How did you embed it? HTML5? I like the embedding itself (versus an embedded video via YT or creating a GIF). 

1. Good catch, thanks! I just updated it!
2. Oh nice, didn't know. I could probably use a matplotlib refresher. Bookmarked!

**Reply by labdmitriy** [reply]
Video is embedded using HTML and rather new [feature](https://github.blog/2021-05-13-video-uploads-available-github/) to upload small video (up to 10Mb) to GitHub:
```html
<video src="https://user-images.githubusercontent.com/9625027/173230769-c069d0f8-8352-4772-8fad-7d6b510d3227.mp4" controls="controls" style="max-width:100%;">
</video>
```
When you drag and drop video file to the issue or PR draft then URL will be returned like in the example above. It’s a small hack but I prefer this alternative now.

**Reply by rasbt** [reply]
Wow nice, had no idea this existed!
